File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>
using namespace std;

bool solution(vector<string> phone_book) {
    map<string, int> hash;
    // make hash
    for(int i=0;i<phone_book.size();i++)
        hash[phone_book[i]] = i+1;
    
    for(int i=0; i<phone_book.size();i++)
    {
        for(int j=0; j<phone_book[i].size()-1;j++)
        {
            string phone_number = phone_book[i].substr(0,j+1);
            //cout << "phone_number: " <<  phone_number << endl;
            //cout << hash[phone_number] ;
            if(hash[phone_number])
                return false;
        }
    }
    return true;
}

// #include <string>
// #include <vector>
// #include <map>
// #include <iostream>
// #include <algorithm>
// using namespace std;

// bool solution(vector<string> phone_book) {
//     map<int, string> hash;
//     // 1. make hash
//     sort(phone_book.begin(), phone_book.end());
//     for(int i=0; i<phone_book.size();i++)
//         hash.insert(make_pair(i, phone_book[i]));

//     char first_char = hash[0][0];
//     int i = 0;
//     while(1)
//     { 
//         if(i == hash.size()-1)
//             break;
//         for(int j=i+1; j<hash.size();j++)
//         {
//             if(first_char == hash[j][0] )
//             {
//                 if(hash[j].find(hash[i]) == 0) // hash[i]를 포함하는지
//                     return false;
//             }
//             else
//             {
//                 first_char = hash[j][0];
//                 i = j;
//             }
//         }
//     }
    
//     return true;
// }

File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.java
import java.io.*;
import java.util.*;

class Solution {
    public boolean solution(String[] phone_book) {
        boolean answer = true;
        Arrays.sort(phone_book,
    (s1, s2) -> 
                    {return Integer.compare(s2.length(), s1.length());}
            
        );
        
        HashMap<String, Integer> hash = new HashMap<>();
        for(String str:phone_book){
            if(hash.containsKey(str)) return false;
            else{
                String tmp ="";
                for(int i=0;i<str.length();i++){
                    tmp += str.charAt(i);
                    hash.put(tmp, 0);
                }
            }




        }
        
        
        return answer;
    }
}